samza-api/src/main/java/org/apache/samza/context/ApplicationContainerContextFactory.java

```diff
@@ -35,14 +36,42 @@
  *
  * @param <T> concrete type of {@link ApplicationContainerContext} created by this factory
  */
+@InterfaceStability.Evolving
 public interface ApplicationContainerContextFactory<T extends ApplicationContainerContext> extends Serializable {
-
   /**
    * Creates an instance of the application-defined {@link ApplicationContainerContext}.
+   * <p>
+   * Applications should implement this to provide a context for container initialization.
    *
+   * @param externalContext external context provided for the application; null if it was not provided
```

Comment:
For consistency: "application-defined external context ...".
Same for all other classes.

Reply:
I had considered this, but since external context is passed through the ApplicationRunner and not through the SamzaApplication (unlike ApplicationContainerContext/ApplicationTaskContext), I didn't want to couple it with being "application-defined".
Do you still think we should add that label?

Reply:
I think the more important contrast is between "framework-provided" and "application-defined", so I'd rather keep it.

Reply:
The example I was thinking about was if someone implemented LocalContainerRunner.buildExternalContext to return some pre-defined set of beans which was not dependent on an application. To me, LocalContainerRunner is "framework" and not "application", so in this case, I felt that the external context was "framework provided".
We will be using the external context abstraction layer as "application-defined", but it seemed like there could be use cases where it falls more under the "framework-provided" category. Maybe we could chat in-person when you have some time.

Reply:
Discussed in-person. We will leave the documentation vague since it could be "framework-provided" or "application-provided" in different cases.